src/waveform/renderers/qopengl/waveformrendererrgb.cpp

```diff
@@ -0,0 +1,297 @@
+#include "waveform/renderers/qopengl/waveformrendererrgb.h"
+
+#include "fixedpointcalc.h"
+#include "track/track.h"
+#include "util/math.h"
+#include "waveform/waveform.h"
+#include "waveform/waveformwidgetfactory.h"
+#include "waveform/widgets/qopengl/waveformwidget.h"
+#include "widget/wskincolor.h"
+#include "widget/wwidget.h"
+
+using namespace qopengl;
+
+WaveformRendererRGB::WaveformRendererRGB(
+        WaveformWidgetRenderer* waveformWidget)
+        : WaveformRendererSignalBase(waveformWidget) {
+}
+
+WaveformRendererRGB::~WaveformRendererRGB() {
+}
+
+void WaveformRendererRGB::onSetup(const QDomNode& node) {
+    Q_UNUSED(node);
+}
+
+void WaveformRendererRGB::initializeGL() {
+    QString vertexShaderCode =
+            "\
+uniform mat4 matrix;\n\
+attribute vec4 position;\n\
+attribute vec3 color;\n\
+varying vec3 vcolor;\n\
+void main()\n\
+{\n\
+    vcolor = color;\n\
+    gl_Position = matrix * position;\n\
+}\n";
+
+    QString fragmentShaderCode =
+            "\
+varying vec3 vcolor;\n\
+void main()\n\
+{\n\
+    gl_FragColor = vec4(vcolor,1.0);\n\
+}\n";
+
+    if (!m_shaderProgram.addShaderFromSourceCode(QOpenGLShader::Vertex, vertexShaderCode)) {
+        return;
+    }
+
+    if (!m_shaderProgram.addShaderFromSourceCode(QOpenGLShader::Fragment, fragmentShaderCode)) {
+        return;
+    }
+
+    if (!m_shaderProgram.link()) {
+        return;
+    }
+
+    if (!m_shaderProgram.bind()) {
+        return;
+    }
+}
+
+inline void WaveformRendererRGB::addRectangle(
+        float x1,
+        float y1,
+        float x2,
+        float y2,
+        float r,
+        float g,
+        float b) {
+    m_lines[m_lineIndex++] = x1;
+    m_lines[m_lineIndex++] = y1;
+
+    m_lines[m_lineIndex++] = x2;
+    m_lines[m_lineIndex++] = y1;
+
+    m_lines[m_lineIndex++] = x1;
+    m_lines[m_lineIndex++] = y2;
+
+    m_lines[m_lineIndex++] = x1;
+    m_lines[m_lineIndex++] = y2;
+
+    m_lines[m_lineIndex++] = x2;
+    m_lines[m_lineIndex++] = y2;
+
+    m_lines[m_lineIndex++] = x2;
+    m_lines[m_lineIndex++] = y1;
+
+    for (int i = 0; i < 6; i++) {
+        m_colors[m_colorIndex++] = r;
+        m_colors[m_colorIndex++] = g;
+        m_colors[m_colorIndex++] = b;
+    }
+}
+
+void WaveformRendererRGB::renderGL() {
+    // The source of our data are uint8_t values (waveformData.filtered.low/mid/high).
+    // We can avoid type conversion by calculating the values needed for drawing
+    // using fixed point. Since the range of the intermediate values is limited, we
+    // can take advantage of this and use a lookup table instead of sqrtf.
```

Comment:
is that optimization really necessary though? It seems premature to me.

Reply:
When I first started looking into the performance issues I thought this code was particularly inefficient and I did this optimization (and I have used it without any issues for a long time). I did some benchmarking and it was definitely faster, but of course the real performance issue was elsewhere. I imagine your concern is that the fixed point math doesn't make it easier to read. I will create a version that uses floats and do another comparison to see if the reduced readability is worth the gain.

Reply:
now using floats